src/main/java/client/Rock.java

```diff
@@ -0,0 +1,150 @@
+package client;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.function.Predicate;
+
+import commands.Commands;
+import io.Invoker;
+import io.Ui;
+import storage.Storage;
+import storage.StorageException;
+import tasks.Task;
+import tasks.TaskList;
+/**
+ * Rock is the name of and the main program used
+ * to run the chatbot for ip.
+ * @author Alvis Ng (supermii2)
+ */
+public class Rock {
+    private static Path filePath = Paths.get("tasks.ser");
+    private TaskList taskList;
+    private Storage storage;
+    private Ui ui;
+    private Invoker invoker;
+    private boolean isTerminated = false;
+    /**
+     * Initialises the chatbot
+     * @param path Path to save and load chat bot data
+     */
+    public Rock(Path path) {
+        this.taskList = new TaskList();
+        this.ui = new Ui();
+        this.invoker = new Invoker(new Commands(this));
+        try {
+            this.storage = new Storage(path.toAbsolutePath().toFile(), this);
+        } catch (StorageException e) {
+            System.out.println(e.getMessage());
+        }
+        this.ui.startup();
+    }
+    private void run() {
+        while (!isTerminated) {
+            String userInput = this.ui.getInput();
+            try {
+                ui.respond(invoker.handle(userInput));
+            } catch (RockException e) {
+                this.ui.respond(e.getMessage());
+            }
+        }
+    }
+    /* Methods for Interactions with task list */
+    /**
+     * Adds the given task to
+     * the task list
+     * @param task Task to be added
+     */
+    public void addTask(Task task) {
```

Comment:
Since all these methods related to TaskList are not directly used in this class, would it be better to put them in the other related classes instead? For example, everything related to TaskList can be put in the TaskList class.

Reply:
Good catch! The methods do already exist in TaskList, but this was an attempt to decouple TaskList from Commands such that I didn't need to pass in the TaskList as an argument or create a getTaskList() method, but I think you're right in that I should just get the TaskList and use methods from there, since manipulation of the TaskList seems like a core part of the majority of commands!